File: generate_book_pdf.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');

const delay = ms => new Promise(r => setTimeout(r, ms));
const BASE_URL = 'http://localhost/chunking-english/book-check.php';
const OUTPUT_DIR = path.join(__dirname, 'pdf_output');
const BATCH_SIZE = 25;

async function makePdf(browser, url, outPath, label) {
    console.log(`  ${label}...`);
    const page = await browser.newPage();
    await page.setViewport({ width: 1200, height: 1600 });
    
    await page.goto(url, { waitUntil: 'networkidle0', timeout: 300000 });
    // 폰트+이미지+텍스트 완전 로딩 대기
    await delay(5000);
    // 모든 이미지가 로딩될 때까지 추가 대기
    await page.evaluate(() => {
        return Promise.all(
            Array.from(document.images)
                .filter(img => !img.complete)
                .map(img => new Promise(resolve => {
                    img.onload = img.onerror = resolve;
                }))
        );
    });
    await delay(2000);

    await page.pdf({
        path: outPath,
        format: 'A4',
        printBackground: true,
        margin: { top: 0, right: 0, bottom: 0, left: 0 },
        timeout: 300000
    });
    
    const sz = fs.statSync(outPath).size / 1024 / 1024;
    console.log(`  ${label} done (${sz.toFixed(1)} MB)`);
    await page.close();
}

async function main() {
    if (!fs.existsSync(OUTPUT_DIR)) fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    // 이전 PDF 정리
    fs.readdirSync(OUTPUT_DIR).filter(f => f.endsWith('.pdf')).forEach(f => fs.unlinkSync(path.join(OUTPUT_DIR, f)));

    const browser = await puppeteer.launch({
        headless: 'new',
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-gpu', '--disable-dev-shm-usage', '--font-render-hinting=none'],
        protocolTimeout: 600000
    });

    // 1. 표지 + 서문
    await makePdf(browser, 
        `${BASE_URL}?mode=cover`,
        path.join(OUTPUT_DIR, 'part_000_cover.pdf'),
        'Cover + Preface');

    // 2. Day 배치 (25일씩)
    for (let from = 1; from <= 250; from += BATCH_SIZE) {
        const to = Math.min(from + BATCH_SIZE - 1, 250);
        const idx = String(Math.ceil(from / BATCH_SIZE)).padStart(3, '0');
        await makePdf(browser,
            `${BASE_URL}?mode=days&from=${from}&to=${to}`,
            path.join(OUTPUT_DIR, `part_${idx}_day${from}-${to}.pdf`),
            `Day ${from}-${to}`);
    }

    // 3. 뒷표지
    await makePdf(browser,
        `${BASE_URL}?mode=last`,
        path.join(OUTPUT_DIR, 'part_999_back.pdf'),
        'Back Cover');

    await browser.close();

    // 결과 요약
    const files = fs.readdirSync(OUTPUT_DIR).filter(f => f.endsWith('.pdf')).sort();
    let total = 0;
    console.log(`\n=== ${files.length} PDF files ===`);
    files.forEach(f => {
        const s = fs.statSync(path.join(OUTPUT_DIR, f)).size;
        total += s;
        console.log(`  ${f} (${(s/1024/1024).toFixed(1)} MB)`);
    });
    console.log(`Total: ${(total/1024/1024).toFixed(1)} MB`);
    console.log('\nNext: python merge_pdfs.py && python compress_pdf.py');
}

main().catch(e => { console.error('Fatal:', e.message); process.exit(1); });

File: generate_book_pdf.py
"""
청킹 영어 전체 PDF 생성기
저작권 등록용 완전본

사용법:
  python generate_book_pdf.py

출력:
  바탕화면/chunking_english_book.pdf
"""

import os, sys, io, tempfile
from pathlib import Path
from PIL import Image
from playwright.sync_api import sync_playwright
import fitz  # PyMuPDF

# ── 설정 ────────────────────────────────────────
LOCAL_URL   = "http://localhost/chunking-english"
BOOK_URL    = f"{LOCAL_URL}/book-check.php"
TOTAL_DAYS  = 250
OUTPUT_PATH = str(Path.home() / "Desktop" / "chunking_english_book.pdf")

# 이미지 압축 설정 (route intercept)
MAX_DIM  = 400   # 최대 이미지 크기 (px)
QUALITY  = 70    # JPEG 품질
# ────────────────────────────────────────────────


def intercept_image(route, request):
    """Playwright route intercept: 이미지를 PDF 삽입 전에 리사이즈/JPEG 변환"""
    try:
        resp = route.fetch()
        if not resp.ok:
            route.continue_()
            return
        body = resp.body()
        pil = Image.open(io.BytesIO(body)).convert('RGB')
        w, h = pil.size
        if max(w, h) > MAX_DIM:
            ratio = MAX_DIM / max(w, h)
            pil = pil.resize((int(w * ratio), int(h * ratio)), Image.LANCZOS)
        buf = io.BytesIO()
        pil.save(buf, "JPEG", quality=QUALITY, optimize=True)
        route.fulfill(
            body=buf.getvalue(),
            content_type="image/jpeg",
            headers={"content-type": "image/jpeg"},
        )
    except Exception:
        route.continue_()


def render_page(page, url, out_path):
    page.route("**/*.png",  intercept_image)
    page.route("**/*.jpg",  intercept_image)
    page.route("**/*.jpeg", intercept_image)

    page.goto(url, timeout=120_000)
    page.wait_for_load_state("domcontentloaded", timeout=120_000)
    try:
        page.wait_for_load_state("networkidle", timeout=30_000)
    except Exception:
        pass
    page.pdf(
        path=out_path,
        format="A4",
        print_background=True,
        margin={"top":"0mm","right":"0mm","bottom":"0mm","left":"0mm"},
        prefer_css_page_size=True,
    )
    doc = fitz.open(out_path)
    n = doc.page_count
    doc.close()

    page.unroute("**/*.png")
    page.unroute("**/*.jpg")
    page.unroute("**/*.jpeg")

    return n


def main():
    print("=" * 55)
    print("  Chunking English - Full Book PDF")
    print("  Copyright Registration Edition")
    print("=" * 55)

    tmp_dir = Path(tempfile.mkdtemp(prefix="chunking_pdf_"))
    print(f"\nTemp dir: {tmp_dir}")

    pdf_parts = []

    with sync_playwright() as p:
        print("\n[1] Browser launching...")
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(viewport={"width": 1280, "height": 900})
        page = context.new_page()

        # 표지
        print("[2] Cover pages...")
        out = str(tmp_dir / "000_cover.pdf")
        n = render_page(page, f"{BOOK_URL}?mode=cover", out)
        print(f"    {n} pages")
        pdf_parts.append(out)

        # Day 1 ~ 250 (1일씩)
        print(f"\n[3] Rendering Day 1 ~ {TOTAL_DAYS}...")
        for day in range(1, TOTAL_DAYS + 1):
            out = str(tmp_dir / f"{day:03d}_day.pdf")
            url = f"{BOOK_URL}?mode=days&from={day}&to={day}"
            try:
                n = render_page(page, url, out)
                print(f"    Day {day:3d}: {n} pages", flush=True)
                pdf_parts.append(out)
            except Exception as e:
                print(f"    Day {day:3d}: ERROR - {e}", flush=True)

        # 서문 / 뒷표지
        print("\n[4] Last page...")
        out = str(tmp_dir / "999_last.pdf")
        n = render_page(page, f"{BOOK_URL}?mode=last", out)
        print(f"    {n} pages")
        pdf_parts.append(out)

        browser.close()

    # PDF 합본
    print(f"\n[5] Merging {len(pdf_parts)} PDFs...")
    merged = fitz.open()
    total_pages = 0
    for path in pdf_parts:
        try:
            doc = fitz.open(path)
            n = doc.page_count
            merged.insert_pdf(doc)
            doc.close()
            total_pages += n
        except Exception as e:
            print(f"    skip {path}: {e}")

    # 임시 파일 정리
    for path in pdf_parts:
        try: os.remove(path)
        except: pass
    try: tmp_dir.rmdir()
    except: pass

    merged.save(OUTPUT_PATH, deflate=True, garbage=4)
    merged.close()

    out_file = Path(OUTPUT_PATH)
    if out_file.exists():
        size_mb = out_file.stat().st_size / (1024 * 1024)
        print(f"\n{'=' * 55}")
        print(f"  DONE!")
        print(f"  File  : {OUTPUT_PATH}")
        print(f"  Pages : {total_pages}")
        print(f"  Size  : {size_mb:.1f} MB")
        print(f"{'=' * 55}\n")
    else:
        print("\n  ERROR: PDF not created.")
        sys.exit(1)


if __name__ == "__main__":
    main()
